### PR TITLE
GH-37929: [Python] begin moving static settings to pyproject.toml

### PR DESCRIPTION
**Pull request by anjakefala**
### Rationale for this change

To migrate Arrow to modern Python packaging standards, see [PEP-517](https://peps.python.org/pep-0517/) and [PEP-518](https://peps.python.org/pep-0518/). 
* GitHub Issue: #37929

This PR focuses on migrating the static settings, the metadata and version, to pyproject.toml. Future PRs will migrate more of the build process to pyproject.toml.

### CODE REVIEW

**Comment by anjakefala** [comment]
I found this! https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata I'll get to work updating.

File: python/setup.py [hunk 1]
```diff
@@ -495,32 +491,10 @@ def has_ext_modules(foo):
     use_scm_version={
         'root': os.path.dirname(setup_dir),
         'parse': parse_git,
-        'write_to': os.path.join(scm_version_write_to_prefix,
+        'version_file': os.path.join(scm_version_write_to_prefix,
                                  'pyarrow/_generated_version.py'),
```
Comment (assignUser) [inline python/setup.py:495]: Looking at the looks this isn't allowed to be an absolute path. Maybe:
```suggestion
        'version_file': 'pyarrow/_generated_version.py',
```

Comment (assignUser) [reply python/setup.py:495]: Yeah, locally that works. I didn't setup with a proper arrow install so cmake fails but should work in ci :)

**Comment by anjakefala** [comment]
This is the current challenge for this PR.

To migrate the static metadata to the pyproject.toml, we need to set a `version`. In PyArrow, the version is set dynamically using setuptools_scm.

Setuptools_scm will only let you [configure Callables in the setup.py](https://setuptools-scm.readthedocs.io/en/stable/config/). There are two callables we set, one for `parse` and one for `version_scheme`.

```
 429 def parse_git(root, **kwargs):                                                                                                                                                                                                
 430     """                                                                                                                                                                                                                       
 431     Parse function for setuptools_scm that ignores tags for non-C++                                                                                                                                                           
 432     subprojects, e.g. apache-arrow-js-XXX tags.                                                                                                                                                                               
 433     """                                                                                                                                                                                                                       
 434     from setuptools_scm.git import parse                                                                                                                                                                                      
 435     kwargs['describe_command'] =\                                                                                                                                                                                             
 436         'git describe --dirty --tags --long --match "apache-arrow-[0-9]*.*"'                                                                                                                                                  
 437     return parse(root, **kwargs)                                                                                                                                                                                              
 438                                          
```

```
 440 def guess_next_dev_version(version):                                                                                                                                                                                          
 441     if version.exact:                                                                                                                                                                                                         
 442         return version.format_with('{tag}')                                                                                                                                                                                   
 443     else:                                                                                                                                                                                                                     
 444         def guess_next_version(tag_version):                                                                                                                                                                                  
 445             return default_version.replace('-SNAPSHOT', '')                                                                                                                                                                   
 446         return version.format_next_version(guess_next_version)                                                                                                                                                                
 447                                                                
```

As they currently are, we cannot configure these in the pyproject.toml, it will not accept a Python callable.

The next part of the challenge is that if you move the version metadata to pyproject.toml, none of the configurations in `setup.py` will be picked up. That is why the build is failing. So you cannot put the static variables in pyproject.toml, and pass the Python callables into setup.py via `use_scm_version`. It is an all-or-nothing migration.

I'm thinking that the next step is to contact the maintainers of setuptools_scm, and see if they have any advice.

**Comment by jorisvandenbossche** [comment]
The setuptools_scm docs have an example of passing a callable in setup.py _with_ using pyproject.toml: https://setuptools-scm.readthedocs.io/en/latest/customizing/#providing-project-local-version-schemes 
So based on that it seems this should be possible?

**Comment by jorisvandenbossche** [comment]
Also, it seems that we use `parse_git` callable to use a custom `git describe` invocation. But, nowadays setuptools_scm also has an option to directly override which describe command is used (`git_describe_command`). So that part can probably be turned into a static configuration instead of a callable.

**Comment by anjakefala** [comment]
>  So based on that it seems this should be possible?

I found open issues for the behaviour I am noticing:

https://github.com/pypa/setuptools_scm/issues/827
https://github.com/pypa/setuptools_scm/issues/1011

**Comment by anjakefala** [comment]
Locally, the test seems to be behaving decently:

```
(base) ~/git/arrow/python kef/pyproject $ ls dist 
pyarrow-16.0.0.dev453+g51a3831e4.d20240416.tar.gz

```

The question is understanding why it is failing in CI.

**Comment by pitrou** [comment]
Note that we're not married to setuptools_scm. If we find out that something else would work better for us, then we can switch to it.

Found this comparison using a quick search: https://github.com/jwodder/versioningit/issues/46#issuecomment-1501201632

**Comment by anjakefala** [comment]
From my investigation, it seems like the problem is that `setuptools-scm` is missing from the release VM:

`   /opt/hostedtoolcache/Python/3.12.3/x64/bin/python3: No module named setuptools_scm`. It's not being installed. I'll look into that.

**Comment by anjakefala** [comment]
Okay, I managed to migrate all of the settings, and it is now working locally and on the CI! This is ready for review.

Note: I'm not quite sure what to do about the macos-latest runner for the "Source Release and Merge Script" build: `Error: The current runner (macos--arm64) was detected as self-hosted because the platform does not match a GitHub-hosted runner image (or that image is deprecated and no longer supported).`

**Comment by assignUser** [comment]
> Note: I'm not quite sure what to do about the macos-latest runner for the "Source Release and Merge Script" build: 

I think this is due to `macos-latest` being moved to an arm64 runner which isn't accounted for in the scripts, ah yes here: #41371

**Comment by assignUser** [comment]
@github-actions crossbow submit -g python

**Comment by anjakefala** [comment]
@assignUser Yeah, this can happen if we install arrow using `python3 setup.py sdist` without having at least 8.0 of `setuptools-scm`. How do I update these environments?

**Comment by anjakefala** [comment]
@github-actions crossbow submit -g python

**Comment by anjakefala** [comment]
We are working on figuring out how to best update these environments. I want to research what is the recommend building method, now that Python is moving away from setup.py. If we need to stick with `python setup.py sdist` I'll need to find each environment that needs setuptools_scm installed.

File: python/pyproject.toml [hunk 2]
```diff
@@ -24,7 +24,53 @@ requires = [
...
+maintainers = [
+    {name = "Apache Arrow Developers", email = "dev@arrow.apache.org"}
+]
+
+[project.urls]
+Homepage = "https://arrow.apache.org/"
+Documentation = "https://arrow.apache.org/docs/python"
+Repository = "https://github.com/apache/arrow"
+Issues = "https://github.com/apache/arrow/issues"
+Changelog = "https://github.com/apache/arrow/blob/main/CHANGELOG.md"
```
Comment (jorisvandenbossche) [inline python/pyproject.toml:59]: This is a very outdated file (which we should maybe remove, or if not keep up to date, but that's for another issue), so I would not link to that. 
(this is not a required entry, so I would just leave it out for now given we don't have a clear one link to the changelog)

File: python/setup.py [hunk 3]
```diff
@@ -421,31 +421,6 @@ def get_outputs(self):
...
-    if version.exact:
-        return version.format_with('{tag}')
-    else:
-        def guess_next_version(tag_version):
-            return default_version.replace('-SNAPSHOT', '')
-        return version.format_next_version(guess_next_version)
-
-
 with open('README.md') as f:
     long_description = f.read()
```
Comment (jorisvandenbossche) [inline python/setup.py:425]: This can be removed as well?

File: python/setup.py [hunk 4]
```diff
@@ -480,47 +455,14 @@ def has_ext_modules(foo):
...
-    use_scm_version={
-        'root': os.path.dirname(setup_dir),
-        'parse': parse_git,
-        'write_to': os.path.join(scm_version_write_to_prefix,
-                                 'pyarrow/_generated_version.py'),
-        'version_scheme': guess_next_dev_version
-    },
-    setup_requires=['setuptools_scm', 'cython >= 0.29.31'] + setup_requires,
-    install_requires=install_requires,
     tests_require=['pytest', 'pandas', 'hypothesis'],
```
Comment (jorisvandenbossche) [inline python/setup.py:466]: I don't know if anyone is relying on this? 
Given one shouldn't run the tests through setuptools (which is also deprecated, I think), I think we can remove this as well (same for `test_suite` below)

We could add a tests "extra" in pyproject.toml with those dependencies if we want (although strictly speaking pandas is not actually a required test dependency, we will skip tests if pandas is not installed. So if we do add this extra, we should use an up to date list of actual test dependencies)

File: python/setup.py (hunk 3 above)
Comment (jorisvandenbossche) [reply python/setup.py:425]: Some other variables that can be cleaned up are `install_requires`, `setup_requires` below, and `default_version` a bit above

**Review by jorisvandenbossche — COMMENTED** [review]
Nice work! Tested it locally in my dev env, and this seems to work fine (after updating setuptools_scm version). 

I got one warning:

```
  /home/joris/conda/envs/arrow-dev/lib/python3.11/site-packages/setuptools_scm/_integration/pyproject_reading.py:59: UserWarning: pyproject.toml: at [tool.setuptools_scm]
  ignoring value relative_to='pyproject.toml' as its always relative to the config file
    warnings.warn(
```

**Comment by jorisvandenbossche** [comment]
> We are working on figuring out how to best update these environments. I want to research what is the recommend building method, now that Python is moving away from setup.py. If we need to stick with `python setup.py sdist` I'll need to find each environment that needs setuptools_scm installed.

In general we should move away from `python setup.py sdist` anyway because those kind of direct setup.py calls are deprecated, so this seems like a good reason to actually do so. 
(see https://setuptools.pypa.io/en/latest/deprecated/commands.html and https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for context)

Specifically for building an sdist or wheel, the recommended way to do that nowadays is using the `build` package (`python -m build --sdist` for sdist, [build.pypa.io](https://build.pypa.io/))

(I know we still document to use `python setup.py build_ext --inplace` for local development, and I also use that myself, for this one there is a less clear alternative)

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

**Comment by raulcd** [comment]
The failures on `example-python-minimal-build-fedora-conda` and `example-python-minimal-build-ubuntu-venv` are due to environmental issues. They are trying to find the tag for `apache-arrow-16.0.0.dev` on the fork and can't compute the version. Those jobs are slightly different than the others as we are using the docker-compose on the minimal examples on python: https://github.com/apache/arrow/tree/main/python/examples/minimal_build
we don't use archery and we don't use the `SETUPTOOLS_SCM_PRETEND_VERSION` variable.
If I execute the jobs locally with:
```
$ cd arrow/python/examples/minimal_build
$ docker-compose build
$ docker-compose run --rm minimal-ubuntu-venv
```
they are successful.
```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==== 6115 passed, 1538 skipped, 7 xfailed, 2 warnings in 79.12s (0:01:19) =====
```
@anjakefala for the sake of investigation could you fetch the tags from the main arrow repo and push them to your fork and retry?
This should fix the crossbow jobs.
If that's the case we should open a follow up issue but this is slightly unrelated to your PR.

**Comment by raulcd** [comment]
When I say, I execute the jobs locally, it is with the changes on your PR :)

File: .github/workflows/dev.yml [hunk 5]
```diff
@@ -120,7 +120,7 @@ jobs:
         shell: bash
         run: |
           gem install test-unit
-          pip install "cython>=0.29.31" setuptools six pytest jira
+          pip install "cython>=0.29.31" setuptools six pytest jira setuptools-scm
```
[the remark this answers is not in the document]
Comment (raulcd) [reply .github/workflows/dev.yml:123]: Tracked here: https://github.com/apache/arrow/issues/41527

File: python/setup.py (hunk 4 above)
Comment (anjakefala) [reply python/setup.py:466]: > So if we do add this extra, we should use an up to date list of actual test dependencies)

So rather than it comprehensively installing all of the dependencies that any test might rely on, you only want the non-optional ones?

**Comment by anjakefala** [comment]
(The force push was just to deal locally with a merge conflict. Any actual update has been committed.)

**Comment by jorisvandenbossche** [comment]
The failures you are seeing might be caused by https://github.com/apache/arrow/pull/41455, which started to do am "out of source" build

File: python/pyproject.toml [hunk 6]
```diff
@@ -24,7 +24,59 @@ requires = [
...
+where = ["."]
+
+[tool.setuptools.package-data]
+pyarrow = ["*.pxd", "*.pyx", "includes/*.pxd"]
+
+[tool.setuptools_scm]
+root = '..'
+version_file = 'pyarrow/_generated_version.py'
+version_scheme = 'guess-next-dev'
+git_describe_command = 'git describe --dirty --tags --long --match "apache-arrow-[0-9]*.*"'
```
Comment (raulcd) [inline python/pyproject.toml:82]: Should we have a fallback version similar to what we had here for out of source builds as we had previously?
https://github.com/apache/arrow/blob/main/python/setup.py#L355-L361
```suggestion
git_describe_command = 'git describe --dirty --tags --long --match "apache-arrow-[0-9]*.*"'
fallback_version = '17.0.0a0'
```
I've tried the above locally and this would fix the existing issues event though we should follow up if this is what we want to do on our builds as @jorisvandenbossche suggested [here](https://github.com/apache/arrow/issues/41429#issuecomment-2118659341): 

We also should update the bump_versions script to update this instead of setup.py and probably the tests for bump_versions (I haven't tested the snippet below but just to give an idea of what I mean):
```diff
diff --git a/dev/release/utils-prepare.sh b/dev/release/utils-prepare.sh
index 5136708..bdf9cd7 100644
--- a/dev/release/utils-prepare.sh
+++ b/dev/release/utils-prepare.sh
@@ -26,10 +26,12 @@ update_versions() {
     release)
       local version=${base_version}
       local r_version=${base_version}
+      local python_version=${base_version}
       ;;
     snapshot)
       local version=${next_version}-SNAPSHOT
       local r_version=${base_version}.9000
+      local python_version=${next_version}a0
       ;;
   esac
   local major_version=${version%%.*}
@@ -110,10 +112,10 @@ update_versions() {
 
   pushd "${ARROW_DIR}/python"
   sed -i.bak -E -e \
-    "s/^default_version = '.+'/default_version = '${version}'/" \
-    setup.py
-  rm -f setup.py.bak
-  git add setup.py
+    "s/^fallback_version = '.+'/fallback_version = '${python_version}'/" \
+    pyproject.toml
+  rm -f pyproject.toml.bak
+  git add pyproject.toml
   sed -i.bak -E -e \
     "s/^set\(PYARROW_VERSION \".+\"\)/set(PYARROW_VERSION \"${version}\")/" \
     CMakeLists.txt
```

Comment (jorisvandenbossche) [reply python/pyproject.toml:82]: Ah, yes, nice catch that there is a `fallback_version` that we can use

(although it is not actually needed for sdists, as the original comment indicated)

**Comment by anjakefala** [comment]
> Specifically for building an sdist or wheel, the recommended way to do that nowadays is using the build package (python -m build --sdist for sdist, build.pypa.io)

Since this invocation requires the presence of an additional dependency (`build`), I've retracted my changes related to it for now. I wanted to make sure other things worked, before I added the complexity of making sure `build` was installed where it needed to be.

**Comment by raulcd** [comment]
The current failures are related to the bump-versions tests I referred on my comment

**Comment by jorisvandenbossche** [comment]
Yeah, I committed your suggestion just to see if that already fixed the main python builds (which it seems it did!)

**Comment by anjakefala** [comment]
I made the suggested changes! https://github.com/apache/arrow/commit/68ce9580318b2d9b534aa23e3d12d3d5cf7d8819

File: dev/release/01-prepare-test.rb [hunk 7]
```diff
@@ -208,10 +208,10 @@ def test_version_pre_tag
         ],
       },
       {
-        path: "python/setup.py",
+        path: "python/pyproject.toml",
         hunks: [
-          ["-default_version = '#{@snapshot_version}'",
-           "+default_version = '#{@release_version}'"],
+          ["-fallback_version = '#{@snapshot_version}'",
```
Comment (raulcd) [inline dev/release/01-prepare-test.rb:213]:
```suggestion
          ["-fallback_version = '#{@release_version}a0'",
```

File: dev/release/post-11-bump-versions-test.rb [hunk 8]
```diff
@@ -172,10 +172,10 @@ def test_version_post_tag
         ],
       },
       {
-        path: "python/setup.py",
+        path: "python/pyproject.toml",
         hunks: [
-          ["-default_version = '#{@snapshot_version}'",
-           "+default_version = '#{@next_snapshot_version}'"],
+          ["-fallback_version = '#{@snapshot_version}'",
+           "+fallback_version = '#{@snapshot_version}a0'"],
```
Comment (raulcd) [inline dev/release/post-11-bump-versions-test.rb:178]:
```suggestion
          ["-fallback_version = '#{@release_version}a0'",
           "+fallback_version = '#{@next_version}a0'"],
```

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build*

**Comment by raulcd** [comment]
- `spark` failures are unrelated and should be fixed if we rebase with the latest main

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build*

File: python/examples/minimal_build/build_conda.sh [hunk 9]
```diff
@@ -97,9 +97,6 @@ export CMAKE_PREFIX_PATH=${ARROW_HOME}${CMAKE_PREFIX_PATH:+:${CMAKE_PREFIX_PATH}
 export PYARROW_BUILD_TYPE=Debug
 export PYARROW_CMAKE_GENERATOR=Ninja
 
-# You can run either "develop" or "build_ext --inplace". Your pick
-
-# python setup.py build_ext --inplace
-python setup.py develop
+python setup.py build_ext --inplace
```
Comment (pitrou) [inline python/examples/minimal_build/build_conda.sh]: We'd really like this to become `pip install -e`, if possible. AFAIU, invoking `setup.py` directly is heavily discouraged nowadays.

Comment (pitrou) [reply python/examples/minimal_build/build_conda.sh]: Oh, I see you changed this while I was commenting :-) Nevermind!

Comment (raulcd) [reply python/examples/minimal_build/build_conda.sh]: I just pushed a change with:
```
# Use the same command that we use on python_build.sh
python -m pip install --no-deps --no-build-isolation -vv .
```

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-ubuntu-venv

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-ubuntu-venv

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-ubuntu-venv

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-ubuntu-venv

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-*

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-ubuntu-venv

File: python/examples/minimal_build/build_venv.sh [hunk 10]
```diff
@@ -68,11 +69,9 @@ export CMAKE_PREFIX_PATH=${ARROW_HOME}${CMAKE_PREFIX_PATH:+:${CMAKE_PREFIX_PATH}
 export PYARROW_BUILD_TYPE=Debug
 export PYARROW_CMAKE_GENERATOR=Ninja
 
-# You can run either "develop" or "build_ext --inplace". Your pick
-
-# python setup.py build_ext --inplace
-python setup.py develop
+# Use the same command that we use on python_build.sh
+python -m pip install --no-deps --no-build-isolation -e .
```
Comment (jorisvandenbossche) [inline python/examples/minimal_build/build_venv.sh:73]: Do you know why the `-e` is needed if then below we actually run tests against installed package? (now, we did install it in editable mode before as well, so fine to just keep this, just wondering why the change to `--pyargs pyarrow` below if we keep this editable)

Comment (jorisvandenbossche) [reply python/examples/minimal_build/build_venv.sh:73]: Maybe also keep the `-v` or `-vv` you had in a previous commit, will be useful for debugging CI issues I think

File: python/pyproject.toml [hunk 11]
```diff
@@ -32,6 +32,11 @@ requires = [
 ]
 build-backend = "setuptools.build_meta"
 
+[options]
+install_requires = [
+    "numpy >= 1.16.6"
+]
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyproject.toml:38]: See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#dependencies-and-requirements for a detailed guide

File: python/examples/minimal_build/build_venv.sh (hunk 10 above)
Comment (raulcd) [reply python/examples/minimal_build/build_venv.sh:73]: I am not sure why the editable mode is required but otherwise it fails importing `pyarrow.lib`. I have been able to validate on my local build tree that without the editable mode the cython compiled files aren't available there, I should investigate further. Those files aren't there without it, I should investigate where are those being installed because they are being generated:
```
python/pyarrow/lib.cpython-310-x86_64-linux-gnu.so
python/pyarrow/lib.cpython-38-x86_64-linux-gnu.so
```
as per the `--pyargs pyarrow` I just copied the same command we use for testing on `python_test.sh` when I was debugging, I can probably revert back this change.

File: python/pyproject.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (raulcd) [reply python/pyproject.toml:38]: Thanks, it seems I don't know how to search :stuck_out_tongue_closed_eyes:

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

File: python/examples/minimal_build/build_venv.sh (hunk 10 above)
Comment (pitrou) [reply python/examples/minimal_build/build_venv.sh:73]: > I am not sure why the editable mode is required but otherwise it fails importing `pyarrow.lib`.

That's probably because the current directory contains the `pyarrow` source directory. If you change the current directory to something else, the error should probably disappear.

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@github-actions crossbow submit example-python-minimal-build-*

File: python/examples/minimal_build/build_venv.sh (hunk 10 above)
Comment (raulcd) [reply python/examples/minimal_build/build_venv.sh]: thanks @pitrou this was a good pointer! I've pushed a change https://github.com/apache/arrow/pull/41041/commits/410f0e0539346b4d5a3aa6b7e284482042e992e8 that removes the editable install requirement

File: python/setup.py (hunk 4 above)
Comment (raulcd) [reply python/setup.py]: I've pushed a change removing those. Will do a final CI run to validate all is good!

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python

File: python/examples/minimal_build/build_venv.sh [hunk 12]
```diff
@@ -35,6 +35,7 @@ source $WORKDIR/venv/bin/activate
 git config --global --add safe.directory $ARROW_ROOT
 
 pip install -r $ARROW_ROOT/python/requirements-build.txt
+pip install wheel
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/examples/minimal_build/build_venv.sh:38]: Ah, I see we actually _do_ list `wheel` as a build requirement ;) But I think nowadays the recommendation is not list `wheel` explicitly. 
While I can't find an explicit reference for this, the original PEP text included `wheel` in the simple example (https://peps.python.org/pep-0518/#build-system-table), but the examples for using setuptools in the packaging guide or on the setuptools documentation (https://setuptools.pypa.io/en/latest/userguide/quickstart.html#basic-use) no longer include `wheel`.

So I _think_ we should remove `wheel` from `build-system.requires`

File: python/pyproject.toml [hunk 13]
```diff
@@ -24,7 +24,63 @@ requires = [
...
+Documentation = "https://arrow.apache.org/docs/python"
+Repository = "https://github.com/apache/arrow"
+Issues = "https://github.com/apache/arrow/issues"
+
+[project.optional-dependencies]
+test = [
+    'pytest',
+    'hypothesis',
+    'cffi',
+    'pytz'
```
Comment (jorisvandenbossche) [inline python/pyproject.toml:68]: The current test dependency extra in setup.py also includes pandas, maybe keep that here as well? Or do we only want the strictly required ones? (which is maybe fine, but then I think pytz can be removed as well?)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

File: dev/tasks/python-minimal-build/github.linux.yml [hunk 14]
```diff
@@ -26,7 +26,7 @@ jobs:
     runs-on: ubuntu-latest
 {{ macros.github_set_env(env) }}
     steps:
-      {{ macros.github_checkout_arrow(submodules=false)|indent }}
+      {{ macros.github_checkout_arrow(fetch_depth=0, submodules=false)|indent }}
```
Comment (anjakefala) [inline dev/tasks/python-minimal-build/github.linux.yml:29]: @raulcd I'm just curious! Why was this change needed?

File: python/pyproject.toml (hunk 13 above)
Comment (anjakefala) [reply python/pyproject.toml:68]: I wasn't sure if we only wanted the strictly required ones! I support adding `pandas`. I thought `pytz` was required, but now I see that it is `importorskip`. =) I'll add `pandas`.

File: dev/tasks/python-minimal-build/github.linux.yml (hunk 14 above)
Comment (assignUser) [reply dev/tasks/python-minimal-build/github.linux.yml:29]: The default fetch depth  for `actions/checkout` is 1. That means it will checkout the specific commit and not fetch anything else, this is useful when dealing with large repos to speed up the checkout but it also leaves out all other branches and more importantly in this case tags (which are needed to calculate the dev version). 0 will fetch the full repo so all tags and branches are available.

Comment (raulcd) [reply dev/tasks/python-minimal-build/github.linux.yml:29]: Yes, that's exactly the case. I realized while testing that the version was not generated correctly due to the missing tags, adding `fetch_depth=0` pulls the tags and generates the correct version for pyarrow.

File: python/examples/minimal_build/build_venv.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (raulcd) [reply python/examples/minimal_build/build_venv.sh]: Done

File: python/examples/minimal_build/Dockerfile.ubuntu [hunk 15]
```diff
@@ -15,7 +15,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
-FROM ubuntu:focal
+FROM ubuntu:jammy
```
[the remark this answers is not in the document]
Comment (raulcd) [reply python/examples/minimal_build/Dockerfile.ubuntu:18]: We are already testing Ubuntu 20.04 on [test-ubuntu-20.04-python-3](https://github.com/ursacomputing/crossbow/actions/runs/9289313780/job/25562914452) .
I did update the minimal_build to 22.04 because there was a minor issue building and was easier to update than investigate to fix it and as we are already testing 20.04 I knew it was a CI set-up. I could investigate but I do think that we have to start updating our default jobs to 22.04 so I preferred to update it.

**Comment by raulcd** [comment]
I've rebased main to fix conflicts

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python

File: python/examples/minimal_build/build_venv.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (raulcd) [reply python/examples/minimal_build/build_venv.sh]: @jorisvandenbossche It seems wheel is required otherwise we get the following failure:
```
 Processing /arrow/python
  Added file:///arrow/python to build tracker '/tmp/pip-req-tracker-grmxa_d3'
  Created temporary directory: /tmp/pip-modern-metadata-sclz_5km
  Preparing metadata (pyproject.toml): started
  Running command Preparing metadata (pyproject.toml)
  <string>:34: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  running dist_info
  creating /tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info
  writing /tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info/PKG-INFO
  writing dependency_links to /tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info/dependency_links.txt
  writing requirements to /tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info/requires.txt
  writing top-level names to /tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info/top_level.txt
  writing manifest file '/tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no previously-included files matching '*.so' found anywhere in distribution
  warning: no previously-included files matching '*.pyc' found anywhere in distribution
  warning: no previously-included files matching '*~' found anywhere in distribution
  warning: no previously-included files matching '#*' found anywhere in distribution
  warning: no previously-included files matching '.DS_Store' found anywhere in distribution
  no previously-included directories found matching '.asv'
  adding license file '../LICENSE.txt'
  adding license file '../NOTICE.txt'
  writing manifest file '/tmp/pip-modern-metadata-sclz_5km/pyarrow.egg-info/SOURCES.txt'
  creating '/tmp/pip-modern-metadata-sclz_5km/pyarrow-17.0.0.dev252+g50a35c1b1.dist-info'
  error: invalid command 'bdist_wheel'
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  ```
  See: https://github.com/ursacomputing/crossbow/actions/runs/9348063637/job/25726407074

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/examples/minimal_build/build_venv.sh]: Ah, OK, diving in a bit more to understand what is happening: this is because we are using `--no-build-isolation` here. So it seems that setuptools will (in normal, isolated mode) automatically inject `wheel` as a dependency when building a wheel, but so that doesn't happen in non-isolated mode.

So it's still a good change to exlude `wheel` from the build requirements in pyproject.toml, but then we need to add it to the environments that use `--no-build-isolation`.

Note that explains this a bit in the setuptools docs (https://setuptools.pypa.io/en/latest/userguide/quickstart.html#basic-use):

>    Historically this documentation has unnecessarily listed ``wheel``
   in the ``requires`` list, and many projects still do that. This is
   not recommended, as the backend no longer requires the ``wheel``
   package, and listing it explicitly causes it to be unnecessarily
   required for source distribution builds.
   You should only include ``wheel`` in ``requires`` if you need to explicitly
   access it during build time (e.g. if your project needs a ``setup.py``
   script that imports ``wheel``).

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit example-python-minimal-build-ubuntu-venv

File: python/examples/minimal_build/build_venv.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply python/examples/minimal_build/build_venv.sh:38]: > So it's still a good change to exlude `wheel` from the build requirements in pyproject.toml,

I may have missed something but is it really useful to exclude it from the build requirements? It's a pure Python package with zero non-trivial dependency, AFAICT.

[the remark this answers is not in the document]
Comment (pitrou) [reply python/examples/minimal_build/build_venv.sh:38]: For example:
```console
$ pip list
Package    Version
---------- -------
pip        24.0
setuptools 59.6.0
$ pip install wheel
Collecting wheel
  Using cached wheel-0.43.0-py3-none-any.whl.metadata (2.2 kB)
Using cached wheel-0.43.0-py3-none-any.whl (65 kB)
Installing collected packages: wheel
Successfully installed wheel-0.43.0
```

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/examples/minimal_build/build_venv.sh:38]: That's the recommendation from setuptools ..
(I agree it shouldn't hurt given it's an easy to install package, although I also noticed in the setuptools tracker that the next version of setuptools will actually remove `wheel` usage all together)

**Review by raulcd — APPROVED** [review]
I think this can be merged now, @jorisvandenbossche @pitrou any other concerns?

**Review by assignUser — APPROVED** [review]

**Review by jorisvandenbossche — APPROVED** [review]
Yes, I would also say, let's merge this ;)

**Comment by anjakefala** [comment]
Thanks everyone! This turned out to be such a surprising doozy.
